modules/cij/runner.py

```diff
@@ -681,80 +739,128 @@ def tcase_enter(trun: TestRun, tsuite: TestSuite, tcase: TestCase) -> int:
         rcode = script_run(trun, hook)
         if rcode:
             break
 
     if trun.conf["VERBOSE"]:
-        cij.emph("rnr:trun::exit { rcode: %r }" % rcode, rcode)
+        cij.emph("rnr:tplan::exit { rcode: %r }" % rcode, rcode)
 
     return rcode
 
 
-def trun_enter(trun: TestRun) -> int:
-    """Triggers when entering the given testrun"""
+def tplan_enter(trun: TestRun, tplan: TestPlan):
+    """Triggers when entering the given testplan"""
 
     if trun.conf["VERBOSE"]:
-        cij.emph("rnr:trun::enter")
+        cij.emph("rnr:tplan::enter")
 
-    trun.stamp["begin"] = int(time.time())     # Record start timestamp
+    tplan.stamp["begin"] = int(time.time())     # Record start timestamp
 
     rcode = 0
-    for hook in trun.hooks["enter"]:     # ENTER-hooks
+    for hook in tplan.hooks["enter"]:     # ENTER-hooks
         rcode = script_run(trun, hook)
         if rcode:
             break
 
     if trun.conf["VERBOSE"]:
-        cij.emph("rnr:trun::enter { rcode: %r }" % rcode, rcode)
+        cij.emph("rnr:tplan::enter { rcode: %r }" % rcode, rcode)
 
     return rcode
 
 
-def trun_setup(conf) -> TestRun:
+def tplan_setup(trun: TestRun, tplan_fpath) -> TestPlan:
     """
-    Setup the testrunner data-structure, embedding the parsed environment
+    Setup the testplan data-structure, embedding the parsed environment
     variables and command-line arguments and continues with setup for
-    testplans, testsuites, and testcases
+    testsuites, and testcases
     """
 
+    tplan = TestPlan()
+
+    tplan.fpath_orig = tplan_fpath
+    tplan.fpath = tplan.fpath_orig
+    tplan.fname = os.path.basename(tplan.fpath)
+    tplan.name = ".".join(tplan.fname.split(".")[:-1])
+    tplan.ident = "%s_%d" % (tplan.name, trun.inc())
+
+    tplan.res_root = os.path.join(trun.conf["OUTPUT"], tplan.ident)
+    tplan.aux_root = os.path.join(tplan.res_root, "_aux")
+
     declr = None
     try:
-        with open(conf["TESTPLAN_FPATH"]) as declr_fd:
+        with open(tplan.fpath) as declr_fd:
             declr = yaml.safe_load(declr_fd)
     except AttributeError as exc:
         cij.err("rnr: %r" % exc)
 
     if not declr:
-        msg = "rnr:trun_setup: failed to read declaration"
+        msg = "rnr:tplan_setup: failed to read declaration"
         cij.err(msg)
         raise InitializationError(msg)
 
+    tplan.evars.update(copy.deepcopy(declr.get("evars", {})))
+
+    os.makedirs(tplan.aux_root)
+
+    # Setup top-level hooks
+    tplan.hooks = hooks_setup(trun, tplan, declr.get("hooks", []))
+
+    for enum, declr in enumerate(declr["testsuites"]):  # Setup testsuites
+        tsuite = tsuite_setup(trun, tplan, declr, enum)
+        tplan.testsuites.append(tsuite)
+
+        trun.progress[Status.Unkn] += len(tsuite.testcases)
+
+    return tplan
+
+
+def trun_exit(trun: TestRun):
+    """Triggers when exiting the given testrun"""
+
+    if trun.conf["VERBOSE"]:
+        cij.emph("rnr:trun:exit")
+
+    return 0
```

Comment:
Should `trun.stamp["end"]` be written here?

Reply:
Time-stamping can't be done in ``on_exit``, since ``on_exit`` might not be invoked in case ``on_enter`` failed. Time-stamping is going on the list of things to improve... so does enter/exit, the logic in the main-runner loop is so convoluted due to this... the enter/exit functions *should* always be invoked, doing things like time-stamping, but hen just not execute hooks in case of enter-failure.